File: src/plugin/AppSearchPlugin.java
package plugin;

import de.jensd.fx.glyphs.materialicons.MaterialIcon;
import gui.objects.icon.Icon;
import gui.objects.textfield.autocomplete.ConfigSearch;
import gui.objects.textfield.autocomplete.ConfigSearch.Entry;
import java.io.File;
import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Stream;
import main.App;
import plugin.Plugin.SimplePlugin;
import util.access.V;
import util.conf.Config.VarList;
import util.conf.Config.VarList.Elements;
import util.conf.IsConfig;
import util.file.Environment;
import util.validation.Constraint;
import static java.util.stream.Collectors.toList;
import static util.file.Util.getFilesR;

public class AppSearchPlugin extends SimplePlugin {

	@Constraint.FileType(Constraint.FileActor.DIRECTORY)
	@IsConfig(name = "Location", group = Plugin.CONFIG_GROUP + ".App Search",
			info = "Root directory the contents of to display "
					+ "This is not a file system browser, and it is not possible to "
					+ "visit parent of this directory."
	)
	final VarList<File> searchDirs = new VarList<>(File.class, Elements.NOT_NULL);

	@IsConfig(name = "Search depth", group = Plugin.CONFIG_GROUP + ".App Search")
	final V<Integer> searchDepth = new V<>(2, this::computeFiles);

	private List<File> apps;
	private final Supplier<Stream<Entry>> src = () -> apps.stream()
			.map(f ->
					ConfigSearch.Entry.of(
							() -> "Run app: " + f.getName(),
							() -> "Runs application: " + f.getAbsolutePath(),
							() -> "Run app: " + f.getAbsolutePath(),
							() -> Environment.runProgram(f),
							() -> new Icon<>(MaterialIcon.APPS)
					)
			);

	public AppSearchPlugin() {
		super("App Search");
		searchDirs.onListInvalid(dirs -> computeFiles());
	}

	@Override
	public void onStart () {
		computeFiles();
		App.APP.search.sources.add(src);
	}

	@Override
	public void onStop () {
		App.APP.search.sources.remove(src);
	}

	void computeFiles() {
		apps = searchDirs.list.stream()
				.distinct()
				.flatMap(dir -> getFilesR(dir, searchDepth.get(), f -> f.getPath().endsWith(".exe")))
				.collect(toList());
	}
}

File: src/plugin/AppSearchPlugin.kt
package plugin

import de.jensd.fx.glyphs.materialicons.MaterialIcon
import gui.objects.icon.Icon
import gui.objects.textfield.autocomplete.ConfigSearch
import main.App
import util.access.v
import util.conf.Config
import util.conf.Config.VarList
import util.conf.Config.VarList.Elements
import util.conf.IsConfig
import util.dev.log
import util.file.Environment
import util.file.nameWithoutExtensionOrRoot
import util.validation.Constraint
import util.validation.Constraint.FileActor.DIRECTORY
import java.io.File
import java.util.function.Supplier

private const val NAME = "App Search"
private const val GROUP = "${Plugin.CONFIG_GROUP}.$NAME"

class AppSearchPlugin: PluginBase(NAME) {

    @Constraint.FileType(DIRECTORY)
    @IsConfig(name = "Location", group = GROUP, info = "Root directories to search through")
    private val searchDirs = VarList(File::class.java, Elements.NOT_NULL)

    @IsConfig(name = "Search depth", group = GROUP)
    private val searchDepth = v(Int.MAX_VALUE)

    @Suppress("unused")
    @IsConfig(name = "Re-scan", group = GROUP)
    private val searchDo = Config.RunnableConfig("rescan", "Rescan apps", GROUP, "", { findApps() })

    private var searchSource: List<File> = emptyList()
    private val searchProvider = Supplier { searchSource.stream().map { it.toRunApplicationEntry() } }

    override fun onStart() {
        findApps()
        App.APP.search.sources.add(searchProvider)
    }

    override fun onStop() {
        App.APP.search.sources.remove(searchProvider)
    }

    private fun findApps() {
        searchSource = searchDirs.list.asSequence()
                .distinct()
                .flatMap { dir -> findApps(dir) }
                .toList()
    }

    private fun findApps(rootDir: File): Sequence<File> {
        return rootDir.walkTopDown()
                .onFail { file, e -> log().warn("Ignoring file={}. No read/access permission", file, e) }
                .maxDepth(searchDepth.value)
                .filter { it.path.endsWith(".exe") }
    }

    private fun File.toRunApplicationEntry() = ConfigSearch.Entry.of(
            { "Run app: $nameWithoutExtensionOrRoot" },
            { "Runs application: $absolutePath" },
            { "Run app: $absolutePath" },
            { Environment.runProgram(this) },
            { Icon(MaterialIcon.APPS) }
    )
}